framework/src/org/checkerframework/framework/type/DefaultInferredTypesApplier.java

```diff
@@ -15,179 +19,111 @@
+        if (inferred == null && primary == null) {
+            // Type doesn't have a primary either, nothing to remove
+            return;
+        } else if (inferred == null && type.getKind() == TypeKind.TYPEVAR) {
+            removePrimaryAnnotationTypeVar(type, inferredTypeMirror, top, primary);
+        } else if (inferred == null) {
+            removePrimaryTypeVarApplyUpperBound(type, inferredTypeMirror, top, primary);
+        } else if (primary != null
+                && (omitSubtypingCheck || hierarchy.isSubtype(inferred, primary))) {
+            type.replaceAnnotation(inferred);
+        } else if (primary == null) {
+            Set<AnnotationMirror> lowerbounds =
+                    AnnotatedTypes.findEffectiveLowerBoundAnnotations(hierarchy, type);
+            AnnotationMirror lowerbound = hierarchy.findAnnotationInHierarchy(lowerbounds, top);
+            if (omitSubtypingCheck | hierarchy.isSubtype(inferred, lowerbound)) {
+                type.replaceAnnotation(inferred);
             }
-
-            return null;
         }
+    }
 
-        /**
-         * For TypeVariables it is important that we both compare the primary annotation and bounds.
-         * Dataflow will default local variables that are also type variables as follows:
-         *
-         *   {@code
-         *       <T> void method() {  @TOP T t; }
-         *   }
-         *
-         * The type @TOP T is equivalent to a type  T[ extends @TOP Object super @TOP Void ]
-         * For the following method:
-         *
-         *   {@code
-         *       <@BOTTOM T extends @TOP Object> void method(T in_t) {
-         *          T t = in_t;
-         *       }
-         *   }
-         *
-         * The type of in_t is the declared type of T and it is below @TOP t, so t should be refined to the
-         * declared type of T[ extends @TOP Object super @BOTTOM Void].  To do this, we must apply the
-         * the annotations of both bounds to type.  This requires a traversal of type.
-         */
-        @Override
-        public Void visitTypevar_Typevar(
-                AnnotatedTypeVariable type, AnnotatedTypeVariable inferred, AnnotationMirror top) {
-            final AnnotationMirror inferredPrimary = inferred.getAnnotationInHierarchy(top);
-            if (inferredPrimary != null) {
-                type.replaceAnnotation(inferredPrimary);
-            } else {
-                type.removeAnnotationInHierarchy(top);
-                visit(type.getUpperBound(), inferred.getUpperBound(), top);
-                visit(type.getLowerBound(), inferred.getLowerBound(), top);
-            }
-            return null;
+    private void removePrimaryTypeVarApplyUpperBound(
+            AnnotatedTypeMirror type,
+            TypeMirror inferredTypeMirror,
+            AnnotationMirror top,
+            AnnotationMirror notInferred) {
+        if (inferredTypeMirror.getKind() != TypeKind.TYPEVAR) {
+            ErrorReporter.errorAbort("Missing annos");
```

Comment:
The error message seems confusing. Shouldn't it be incorrect type kind? Add the type to the error message.